core/src/upgrade.rs

```diff
@@ -152,3 +154,22 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
     /// The `info` is the identifier of the protocol, as produced by `protocol_info`.
     fn upgrade_outbound(self, socket: T, info: Self::Info) -> Self::Future;
 }
+
+/// Possible security upgrade on a connection
+pub trait SecurityUpgrade<T>: UpgradeInfo {
```

Comment:
Can you implement this in `libp2p-tls` and `libp2p-noise` please?

Reply:
I think we may also have to split this one into two traits, otherwise the concrete security upgrade doesn't know, which role it is in (dialer vs listener).

Reply:
Yes, after removing the panic, now we cannot differentiate based on the variant of the `PeerId` parameter.

Reply:
I implemented the new traits in `libp2p-tls` and `libp2p-noise`.